src/crimson/osd/pg.cc

```diff
@@ -807,6 +807,34 @@ PG::get_or_load_clone_obc(hobject_t oid, ObjectContextRef head)
   }
 }
 
+template<RWState::State State>
+seastar::future<>
+PG::with_head_obc(hobject_t oid, with_obc_func_t&& func)
+{
+  assert(oid.is_head());
+  auto [obc, existed] = shard_services.obc_registry.get_cached_obc(oid);
+  auto loaded = load_obc_ertr::make_ready_future<>();
+  return obc->with_lock<State>(
+    [oid, existed=existed, obc=std::move(obc), func=std::move(func), this] {
+    auto loaded = seastar::make_ready_future<ObjectContextRef>(obc);
+    if (existed) {
+      logger().debug("with_head_obc: found {} in cache", oid);
+    } else {
+      logger().debug("with_head_obc: cache miss on {}", oid);
+      loaded = obc->with_promoted_lock<RWState::RWEXCL>([this, obc] {
+        return load_head_obc(obc);
+      });
```

Comment:
If I understand correctly, this means that, with this pr, dropping excl lock of obc and locking it with another type are separated in different continuations, which is different from our current approach in which we call degrade_excl_to() to get a lock on a just-loaded obc. I'm not saying this is problematic, I just feel that this might be worth noting:-)

Reply:
>  in which we call degrade_excl_to() to get a lock on a just-loaded obc.

true. it's like

```c++
auto obc = lrc_cache.get(oid);
// assume obc did not exist in cache
// ready future
return obc.lock<read>().then([&] {
  // promote the lock
  // ready future
  obc.promote_lock<read, exclusive>();
}).then([&] {
  return pg.load_meta(obc);
}).then([&](auto meta) {
  // ready future
  obc.set_meta(meta);
}).then([&] {
  // ready future
  obc.demote_lock<exclusive, read>();
}).then([&] {
  return func(obc);
}).then([&] {
  // ready future
  obc.unlock<read>();
});
```

so i think `obc.demote_lock<exclusive, read>()` can be inlined. so practically, it's the same as what we have before the change.